### ISSUE TITLE
Installation instructions Raspberry - Bluetooth

### ISSUE THREAD

**Issue by wtal710174** [WouterJD/FortiusANT#192]
The intend is to create installation instructions for RaspberryPi OS. Since I have a RaspberryPi 3B this will be mostly base on this device. Let's exchange here to make this an instruction for all Raspberries around. 

I am running a Fortius Trainer (T1930) with a T1932 head unit and the T1941 motor combined with the VR Steering Frame.

My original instruction can be found here: https://gist.github.com/wtal710174/23081b986ab05c9a17facfec2d738202. However I would like to continue the work in this issue.

Using the instructions I have the Raspberry control the Fortius Trainer with only the USB cable connected to it. I run Zwift on my iPad or Laptop connected using Bluetooth. I run the FortiusANT without the GUI in a terminal.

# FortiusANT BLE on Raspberry PI 3B Installation

This is a short description of the necessary steps for the installation of the FortiusANT project by [WouterJD](https://github.com/WouterJD). The project is located here (https://github.com/WouterJD/FortiusANT).
The first part of this document is taken from the USB ANT Installation Instruction written by [martin-vi](https://gist.github.com/martin-vi) which is located here (https://gist.github.com/martin-vi/5206b6b26faf6f2d6121f6c8f300e44d).

## Prerequisits

- Raspberry Pi 3b or similar with a clean RaspiOS Buster full installation (https://www.raspberrypi.org/software/)
- Python3 version > 3.6 (Version 3.7.3 is included in the Buster full installation December 2nd 2020)
- Standard user pi or any other user which belongs to the groups sudo, ssh, dialout, audio, video, plugdev, users, input, netdev (not sure if you need all of these, however this worked for me)
- Enable ssh and/or VNC in the Raspberry Pi Configuration in Preferences (Interfaces tab), if you want to connect remotely to your Raspberry Pi
- Update your system (did not work for me during the initial setup, make sure that you run the apt update command at least before installing any other packages, running both commands does not hurt, just takes some time):
```
$ sudo apt update
$ sudo apt full-upgrade
```
- FortiusANT repository downloaded to the user home directory
via git:
```
$ sudo apt install git
$ git clone https://github.com/WouterJD/FortiusANT.git
```
or via download
```
$ curl -L https://github.com/WouterJD/FortiusANT/archive/master.tar.gz | tar -xzvf -
```
Now you should have a FortiusAnt folder in your home. Check with ls -all if the folder belongs to you.

_There is an updated version of this instruction [here](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-766039238), which introduces some improvements. You can continue with the updated version from this point. Just click on the link._

## Install wxPython

Download pre-build [wxPython package](https://drive.google.com/drive/folders/1owWkRTsvcnvq7ZSRc6zHS5dip2AMvW79?usp=sharing) to home folder. This wheel package works for an ARM7l CPU.

Install package with
```
$ sudo pip3 install wxPython-4.1.1-cp37-cp37m-linux_armv7l.whl
```
If you can not use the prebuild package, you need follow the install instructions below to build your own wxPython by replacing the version with 4.1.1 (This can take up to 18 hours depending on the RaspberryPi you are using):

https://www.marcdobler.com/2020/05/17/how-to-compile-and-install-wxpython-on-raspberry-pi/
## Install required packages
```
$ sudo apt install -y python-virtualenv python3-virtualenv build-essential bluetooth bluez libbluetooth-dev libudev-dev
$ sudo apt-get install -y libatlas-base-dev # for numpy, https://github.com/numpy/numpy/issues/11110
```
## Setup virtual environment for Python
```
$ cd ~
$ virtualenv -p python3 --system-site-packages venv
```
Afterwards, the installation of packages for FortiusANT always requires the virtualv to be activated first. This also applies to starting FortiusANT.

The virtual environment needs to be activated for the following with this command:
```
$ source ~/venv/bin/activate
```
## Node JS installation
Follow the installation instruction on https://www.makersupplies.sg/blogs/tutorials/how-to-install-node-js-and-npm-on-the-raspberry-pi

For the Raspberry Pi 3B the following steps apply for version 14.15.3:
```
# Go to home
$ cd ~
# Download node 14.15.3
$ wget https://nodejs.org/dist/v14.15.3/node-v14.15.3-linux-armv7l.tar.xz
# Untar downloaded archive
$ tar xf node-v14.15.3-linux-armv7l.tar.xz
# Go into folder with unzipped files
$ cd node-v14.15.3-linux-armv7l/
# Copy folder content to 
$ sudo cp -R * /usr/local/
# Check if installed correctly
$ node -v
# Output: v14.15.3
$ npm -v
# Output: 6.14.9
```
After installing node to the system, you have to set it up for FortiusANT
```
$ cd ~/FortiusANT/node/
$ npm install
```
The npm install command will display several warnings which can be ignored.

Last step is to grant the node binary cap_net_raw privileges, so it can start/stop BLE advertising without root access.
```
$ sudo setcap cap_net_raw+eip $(eval readlink -f `which node`)
```
## Install FortiusANT dependencies
```
# remember that an activated virtualenvironment is required
$ pip3 install -r ~/FortiusANT/pythoncode/requirements.txt  
``` 
## Create startup script for FortiusANT
_Check the updated script for starting without sudo in the comment 'Bluetooth without Sudo and updated script' below_
```
$ cd ~
# Deactivate virtual environment
$ deactivate
$ touch FortiusANT.sh
# Open the new file with any editor you like, I prefer pico or nano
$ pico FortiusANT.sh
# Insert the following text including the !/bin/bash comment

#!/bin/bash

echo "Enter the sudo password, please."
read -s PW
echo Password entered. Thank You!

echo $PW|sudo -S service bluetooth stop
echo Bluetooth Service stopped
source /home/pi/venv/bin/activate
if [[ "$VIRTUAL_ENV" != "" ]]
then
  echo Virtual Environment set
else
  echo Virtual Environment not set
fi
echo Virtual Environment Set
echo $PW|sudo -S hciconfig hci0 up
echo Bluetooth for BLE on internal adapter started
cd /home/pi/FortiusANT/pythoncode
echo Starting FortiusANT ...
# Use any command line parameter as required. '-b' does start the command line based bluetooth interface 
# only, without GUI. Add '-g' for the GUI. More Details in the FortiusANTUserManual in section 4.2 
# starting on page 29. 
python3 FortiusAnt.py $@
deactivate
if [[ "$VIRTUAL_ENV" != "" ]]
then
  echo Virtual Environment set
else
  echo Virtual Environment not Set
fi
echo $PW|sudo -S service bluetooth start
echo Standard Bluetooth Service restarted
```
Save file with control+O and exit file with control+x and make script executable with the following command
```
$ chmod u+x FortiusANT.sh
```
## Give user access to USB port
Run the following commands to create an access file in /etc/udev/rules and reboot Raspberry to activate rule
```
$ sudo -i
$ cat << EOF > /etc/udev/rules.d/10-usbaccess.rules
# Allow users in group usbtacx to access usb
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", MODE="0664", GROUP="usbtacx"
EOF
# Create group usbtacx
$ addgroup usbtacx
# add user pi to this group
$ adduser pi usbtacx
# Reboot to activate rule
$ reboot now
```
## Let's fire it up
Connect your trainer to one of the USB ports of the Raspberry Pi and execute FortiusANT with
```
# -b for command line bluetooth startup, add any command line parameter as required, like -g to start with GUI
$ ~/FortiusANT.sh -b
```
Next, the calibration procedure will start with a pedal kick. After the procedure is complete, the bluetooth starts broadcasting identifying the Raspberry Pi as periphal device with the identifier 'FortiusANT Trainer'. This one you can connect to your CTP.

After the calibration you might see a couple of error messages. This happens while the http server is being established and the program is not able to access it. After 5-10 secondes the connection should be stable and you will see the normal output in the terminal like the one below.

```
20:08:31,698: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:32,700: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:33,702: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:34,703: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:35,705: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:36,706: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
```

**Comment by switchabl** [comment]
Don't have a Raspberry PI 3B so can't test this directly, but it looks quite thorough! I have a few suggestions on how to improve this further (well, I hope so at least):
- I suggest putting the venv inside the FortiusANT directory, not directly in home
- Maybe node.js can be installed locally as well (instead of system-wide in /usr/local) with something like nvm (similar to venv). I think with nodeenv it might even be possible to install node directly inside the python venv.
- Passing through additional command line options from the script to FortiusANT would probably be nice.
- I don't think you need to activate the venv in the script, it should be enough to run `~/venv/bin/python FortiusAnt.py`.
- I wonder if hciconfig is actually necessary? I thought that the system Bluetooth stack is actually bypassed completely with the current implementation but I have to admit I didn't look at it in detail.
- The USB permissions could be made a bit narrower (similar to https://github.com/WouterJD/FortiusANT/issues/176#issuecomment-749847993), but on a Raspberry PI it is maybe not so critical I suppose.

Some more general ideas that may or may not make sense to persue:
- A simplified no-GUI version of the guide. I think if you run without GUI, you don't actually need wxPython anymore. So it might be sufficient to install the numpy, requests and pyusb packages with apt and do away with venv.
- It should be possible to give the pi user permission to start/stop Bluetooth with polkit, but I would need to look up how. That way it might be possible to run the whole script without sudo (and storing the password, I am not sure "read -s" is actually secure).
- Alternatively we could run FortiusANT as a systemd service and with a dedicated user. That won't work with a GUI, but it could be particularly attractive if you use a dedicated, headless Raspberry PI for your trainer. Then you might just set the the service to auto-start on boot. I think if you declare a conflict with the bluetooth service in the unit file it will even auto-stop that for you.
- I wonder if we could just make a docker container to save people the hassle of setting it up manually.

**Comment by switchabl** [comment]
> Am I correct that I won’t need a Ant dongle anymore?

If the device you use Zwift/Rouvy/etc on has Bluetooth LE (or if you are using Zwift on PC with the Companion app on the phone) as well and you are not using any ANT-only power meter/sensor, yes, no ANT dongle required.

> I’m willing to share an image if I succeed. That should work for other users with the same RPi then right?

I think so, yes. There are a number of potential disadvantages to building images that way:
- they may not work on other Raspberry PI-models
- if you expanded the partition to the whole SD card, the image may become quite large (and can only installed on cards at least as big as yours)
- they may contain unnecessary (temp/log) files
- it is relatively time-consuming if you wanted to update it regularly

For that reason, images are usually built in an automated way on a PC/server instead. But that is quite a bit more work to set-up, so unless someone volunteers to do that, I think your image could still be quite helpful for others!

Not sure what the best options are long-term:
- distributing normal Debian/rpm packages and relying on the OS for dependencies would be the best solution, but that may not be feasable anymore with the node.js dependencies for Bluetooth
- an automatically built Linux image would be great for people using a dedicated Raspberry PI, but I am not sure how much work that is
- a Docker container would be more flexible and probably easier to build
- there is also AppImage, which would be similar to the binaries provided on Windows (that should also work with the GUI)

**Comment by switchabl** [comment]
Not sure what the best options are long-term:
- distributing normal Debian/rpm packages and relying on the OS for dependencies would be the best solution, but that may not be feasable anymore with the node.js dependencies for Bluetooth
- an automatically built Linux image would be great for people using a dedicated Raspberry PI, but I am not sure how much work that is
- a Docker container would be more flexible and probably easier to build
- there is also AppImage, which would be similar to the binaries provided on Windows (that should also work with the GUI)

**Comment by wouterstrand** [comment]
I'm stuck at the following step:


_The virtual environment needs to be activated for the following with this command:_

`$ source ~/venv/bin/activate`

I get the message the file or folder does not exist? Using command ls -all, it does not exist indeed in the home folder?

**Comment by wtal710174** [comment]
Thanks @switchabl very good and helpful comments. Here my two cents:
- I will try and test putting venv in the FortiusANT directory, that should be an easy change
- 

> 
> 
> I'm stuck at the following step:
> 
> _The virtual environment needs to be activated for the following with this command:_
> 
> `$ source ~/venv/bin/activate`
> 
> I get the message the file or folder does not exist? Using command ls -all, it does not exist indeed in the home folder?

test "$ source --help", does this give you the help messages? If not something went wrong with your virtual environment install.

**Comment by wouterstrand** [comment]
Wooops, noob at work here. I skipped a step, I've succeeded now. Continuing install on the RPi3.

**Comment by wtal710174** [comment]
@switchabl Thanks a lot for the great comments. I will pick this up and try to work through your proposal. I do agree with almost all of it. I did try to run without stopping the bluetooth service and not using the hciconfig command, however this did not work for me. Maybe you have a suggestion there. Regarding the virtual environment in the script, this was just to make the startup as easy as possible. I have already altered the text to include the command line parameters for startup. Very nice suggestion there.

Regarding your general ideas:
- Non Gui version:  Yes this is a great idea. This will require changes in the requirement.txt files to prevent the check for the wxPython or a completely altered program version without the GUI calls. Not sure which option to choose. I prefer the changes in the Python to detect if wxPython is present and in case it is not ignore all the GUI options.
- It should be possible to give the pi user permission to start/stop Bluetooth with polkit: Not requiring sudo would be perfect. I have no experience with polkit. I will look it up and start working on this.
- Alternatively we could run FortiusANT as a systemd service: This would be my favorite solution, since it would allow me to power up the Raspi and start cycling without logging in and typing any command. I do think that you will still need the calibration process. Keeping the FortiusANT on Raspi running all the time seems to cause issues over time. I need to look into that more closely.
- Docker: YES! Unfortunately I have no experience at all with the docker concept. Therefore I could use some help to get this done. This would be a very convenient way of setup. I would as well prefer this to an image.

**Comment by wouterstrand** [comment]
Well, Thanks a bunch. I successfully connected my iPhone to the Tacx via BLE. Tomorrow I will set my bike up on the Tacx and make an attempt with my laptop.

**Comment by wtal710174** [comment]
Nice! Thanks for the feedback. Good luck with the test tomorrow.

**Comment by switchabl** [comment]
> I did try to run without stopping the bluetooth service and not using the hciconfig command, however this did not work for me. Maybe you have a suggestion there.

It seems I was wrong, I looked at the bleno docs (the library uesd for the Bluetooth functionality) and apparently they are both necessary.
But it should be possible to set up permissions so we don't need root for them:
- `hciconfig`: According to https://unix.stackexchange.com/questions/96106/bluetooth-le-scan-as-non-root, this should do the trick:
```
sudo setcap 'cap_net_raw,cap_net_admin+eip' `which hciconfig`
```
- bluetooth service: add a polkit rule (see https://wiki.archlinux.org/index.php/Polkit#Allow_management_of_individual_systemd_units_by_regular_users, replace wpa_supplicant with bluetooth)

> Regarding the virtual environment in the script, this was just to make the startup as easy as possible.

Yes, if you are using it interactively, it definitely makes sense. In a script it apparently isn't necessary most of the times as long as you call the right Python interpreter. If it is (in case FortiusANT expects the Python interpreter in PATH or something), I believe you can at least skip the `deactivate` step. The changes to the environment should revert once the script terminates.

>     * Non Gui version:  Yes this is a great idea. This will require changes in the requirement.txt files to prevent the check for the wxPython or a completely altered program version without the GUI calls. Not sure which option to choose. I prefer the changes in the Python to detect if wxPython is present and in case it is not ignore all the GUI options.

I was under the impression that the latest version of FortiusANT doesn't import wx in CLI mode. So I was hoping that
```
sudo apt install git python3-usb python3-numpy python3-requests
python3 pythoncode/FortiusAnt.py -a
```
may be all that is required (apart from the permissions and Bluetooth part). But maybe not, I haven't tried yet.

>     * Alternatively we could run FortiusANT as a systemd service: This would be my favorite solution, since it would allow me to power up the Raspi and start cycling without logging in and typing any command. I do think that you will still need the calibration process. Keeping the FortiusANT on Raspi running all the time seems to cause issues over time. I need to look into that more closely.
> 

Yes, I wouldn't be surprised if running it continously uncovered some stability bugs.

I have some plans to improve calibration (save calibration value between sessions; allow calibration on request, not on startup; allow calibration requests through ANT/Bluetooth) that would probably help with the head-less case as well.

**Comment by TacxBiker** [comment]
Hi, by excecuting all the commands wtal710174 it sure works perfectly on a raspberry 3b+
Also the GUI works

Adding 
/home/pi/FortiusANT.sh -a -b
to the /etc/rc.local starts up FortiusANT. 
One pedal kick and the calibration starts. After calibration the FortiusANT trainer is visible by bluetooth on a PC or iPAd

**Comment by wouterstrand** [comment]
It seems as if there is no calibration for me? It just immediately outputs data and transmits BLE. Is it because I use the magnetic T190 brake with T1932 head unit?

**Comment by mikeherriman** [comment]
I can also attest that the natallation and configuration steps work on a raspberry Pi 4.  I have a 7” touchscreen connected to the raspberry Pi HDMI port allowing for startup/shutdown without external keyboard or mouse.  Happily able to run Zwift app on iPhone, Zwift app on Windows with FortiusAnt connected via iPhone Zwift Companion app.

There does seem to be some GUI related issues, in that the GUI becomes unresponsive, and requires a kill command to close out.  Even though the GUI becomes unresponsive, FortiusANT continues to function.

Mike

**Comment by wouterstrand** [comment]
How do I start the GUI? I attempted with -g but it didn’t work.

**Comment by TacxBiker** [comment]
It is also possible to start FortiusANT by simply clicking in the script /home/pi/FortiusANT.sh in the filebrower.
Then the GUI is visible.

**Comment by wouterstrand** [comment]
Thanks!

Outlook voor iOS<https://aka.ms/o0ukef> downloaden
________________________________
Van: TacxBiker <notifications@github.com>
Verzonden: Thursday, January 7, 2021 8:39:59 AM
Aan: WouterJD/FortiusANT <FortiusANT@noreply.github.com>
CC: wouterstrand <wouter_reinhard@hotmail.com>; Comment <comment@noreply.github.com>
Onderwerp: Re: [WouterJD/FortiusANT] Installation instructions Raspberry - Bluetooth (#192)


It is also possible to start FortiusANT by simply clicking in the script /home/pi/FortiusANT.sh in the filebrower.
Then the GUI is visible.

**Comment by TacxBiker** [comment]
One thing, it is not possible to add opperants when clicking on the script.
I hust eddited the scrip by addeing te opperants -a -b -g

**Comment by wtal710174** [comment]
I created a little add on which might be useful to others as well. When I started the GUI today for testing some stuff I noticed that the GUI is showing new versions in the status bar if available (very nice feature :-)). Therefore I implemented an automatic update in the script, to check for the latest version during start up and update the new files. In order to do this you need to implement the following.

Step 1:
```
$ cd ~/FortiusANT/
$ git remote add upstream https://github.com/WouterJD/FortiusANT
```
Step 2 - Get new code from repository (needs to be done in the FortiusANT directory)
```
$ git pull upstream master
```
If you implement this line in the script after the password section with
```
git -C /home/pi/FortiusANT pull upstream master
```
the script will always check for updates and download the files before starting FortiusANT.

**Comment by wtal710174** [comment]
# Bluetooth without sudo for pi

I followed @switchabl 's advice. This worked kind of for me. So step 1 is complete ... pi can start and stop bluetooth, however a password is still required. Did anybody get this working without password request for pi so far?

**Comment by wtal710174** [comment]
> One thing, it is not possible to add opperants when clicking on the script.
> I just eddited the scrip by addeing te opperants -a -b -g

@TacxBiker Sounds like you do not need a sudo password for the script anymore, how did you get this done? Or are you starting the script in a terminal after double clicking?

**Comment by wtal710174** [comment]
On more idea for improvement. Some of the users here talk about running the Pi headless and just attaching it to the trainer to make it kind of an interface that does not require any input through a terminal or touchscreen. @switchabl  mentioned the service before. What about connecting an on/off switch to the IO-ports of the raspberry, which starts the script or service in on position and stops the script or service in off position. Anybody that did something similar in the past and has an idea how to implement this for the user pi (script) or as a service that listens to the switch status?

**Comment by wtal710174** [comment]
@mikeherriman Thanks for the feedback!

**Comment by wtal710174** [comment]
### Bluetooth without Sudo and updated script

_Based on the remarks from @switchabl above._ 

First give user the right to use hciconfig without sudo

```
$ sudo setcap 'cap_net_raw,cap_net_admin+eip' `which hciconfig`
```
Then enter the command 'sudo visudo' to edit the sudoers file. In the section 
'# Allow members of group sudo to execute any command' add the following lines:
```
%usbtacx ALL=NOPASSWD: /bin/systemctl stop bluetooth.service
%usbtacx ALL=NOPASSWD: /bin/systemctl start bluetooth.service
```
This will grant all member of the group usbtacx the right to start bluetooth service without entering a password.
Now you can use `sudo service bluetooth stop` and `sudo service bluetooth start` as user pi without entering a password.

_(Comment: I tried to run the command without sudo, since this works for the hciconfig command. If you ommit the sudo you will have to enter a password)_

Now edit the FortiusANT script with `$ pico ~/FortiusANT.sh` and replace the content with the following code:
```
# FortiusANT Startup Script v0.2

# Remove comment below if you want automatic updates
# Execute 'git remote add upstream https://github.com/WouterJD/FortiusANT' in FortiusANT directory first before removing comment
#echo Getting FortiusANT updates from Github Repository
#git -C /home/pi/FortiusANT pull upstream master

sudo service bluetooth stop
echo Bluetooth Service stopped
source /home/pi/venv/bin/activate
if [[ "$VIRTUAL_ENV" != "" ]]
then
  echo Virtual Environment set
else
  echo Virtual Environment not set
fi
hciconfig hci0 up
echo Bluetooth for BLE on internal adapter started
cd /home/pi/FortiusANT/pythoncode
echo Starting FortiusANT ...
python3 FortiusAnt.py $@
sudo service bluetooth start
echo Standard Bluetooth Service restarted
```
Save the changed file and give it a try. Let me know if this works.

Now you should be able to start the script without any issue by double clicking the file in the file browser in the graphical interface or as before in a terminal with `~/FortiusANT.sh -b` to just launch the bluetooth interface.

**Comment by mikeherriman** [comment]
I just updated the sudouser file as instructed, edited the FortiusANT.sh file as instructed.  I also added the switches inside the script (-a -g -H0 -c12).

Created desktop shortcut to the script.
Now can double-click the shortcut from Desktop and everything fires up.  
One thing that I am not sure about is when I double click the shortcut, I am prompted to Execute, Execute In Terminal, Cancel.
When I select Execute in Terminal, everything goes but within a minute or so, the GUI and terminal window become unresponsive, requiring CTRL+C to stop the script.  When I select Execute the GUI starts with no background terminal window and all seems to operate much better.  Any ideas?

**Comment by TacxBiker** [comment]
@wtal710174 
I'm not sure why that is. when auto start the script by RC.local I don't need to enter a password.
when using the same script by clicking or command line in a terminal I do need to enter a password.

**Comment by wtal710174** [comment]
@TacxBiker Which script are you using? Please make sure you have `#!/bin/bash`in the first line. I forgot to add this yesterday, which caused the issues I was having.
Regarding RC.local startup should launch the script as root, which could explain the behavior. You can double check by logging the output of the script in a text file using `whoami` to print the user name when executing the script.  
I have updated the script in the comment above to include the reference to the bash.

**Comment by wtal710174** [comment]
### FortiusAnt One Button Controller for Raspberry

I have created a little python program (actually my first one ever) to control the headless Raspberry with a very simplistic interface (1 Button and a traffic light LED Module). See picture below.

If implemented in the .bashrc of the user pi it will start when the  raspberry is booted and indicate the status using the 3 LED's.

This is a quick and dirty first build, however it makes the control much easier and I do not need to use a keyboard or terminal anymore to stop and start or shutdown.

GPIO Pins are:
- GPIO 24 for the switch input
- Pin 17 for the 3.3V power to the switch
- RED LED on GPIO 22 (PIN 15)
- YELLOW LED on GPIO 27 (PIN 13)
- GREEN LED on GPIO 17 (PIN 11)
- GROUND is connected to PIN 9

If this is of any interest to anybody else, I would be interested in a more professional and rigid module to be connected to the GPIO pins.

Feedback and help to improve the code appreciated.  

![IMG_7486](https://user-images.githubusercontent.com/62606555/104126968-c5c5b880-535f-11eb-936d-6fb7bbd25c96.JPG)
![6E2B9350-B0BB-4489-9E38-973DB92ACE63_1_105_c](https://user-images.githubusercontent.com/62606555/104126987-d5450180-535f-11eb-90df-3cfa0a9bb424.jpeg)
![FortiusAntControl Trafic Light](https://user-images.githubusercontent.com/62606555/104127012-f3aafd00-535f-11eb-9d4d-1f890b2925f4.png)

[FortiusAntControl.py.zip](https://github.com/WouterJD/FortiusANT/files/5792800/FortiusAntControl.py.zip)

**Comment by MeanHat** [comment]
Hi @wtal710174 - looks very good. I was planning to use an illuminated LED switch such as this https://thepihut.com/products/rugged-metal-on-off-switch-with-red-led-ring?_pos=3&_sid=0f4c526eb&_ss=r to shutdown the RPi Zero. At present my neighbour uses a keyboard to command a poweroff/shutdown.
![IMG-20210108-WA0005](https://user-images.githubusercontent.com/73789300/104131544-d7fb2300-536e-11eb-9e2a-e68f2485df4f.jpg)

**Comment by wtal710174** [comment]
@MeanHat Thanks for sharing this. I like the button idea. Looks very nice. Maybe I can find a Case with GPIO Connector that is big enough to fit this button. However I checked on the website and I think I like this switch even better:
https://thepihut.com/products/rugged-metal-pushbutton-16mm-6v-rgb-momentary
Check it out. It has a RBG LED ring, which allows you to show the status of the process, similar to what I am doing with the traffic light. Letting the RED LED ring blink in different ways could to the same.

**Comment by wtal710174** [comment]
I am having some stability issues with the nodejs server. The server process just disappears after a random time (3-6 minutes) while riding. The CPU load is at 12% max when this happens and the memory is about 25% used. So I do not see any resource issue. Does anyone have an idea? I will check the repository if I can find it in another issue. 
Interesting is that this happens when the FortiusAnt.py is started using the FortiusAntController.py program. The FortiusAnt program continues to operate (just no Bluetooth signal anymore). If I am starting the FortiusAnt program using my FortiusANT.sh bash script, this does not happen.

**Comment by MeanHat** [comment]
Hi @wtal710174 
I now have my RPi 3B up and running in BLE mode and communicating with an Android phone running Zwift. I am short of a head unit but borrowed my neighbour's T1932 and it seemed to work OK (though I was not able to check the commands from Zwift were reaching the trainer. Brilliant - thank you!
By way of feedback (1) using FortiusANT.sh I received a "too many arguments" error when I entered 4 arguments (-a -b -d127 -s) and (2) with -b I received an error ("expected binary...).
I had previously set up a RPi Zero using system.d to automatically start FortiusANT which worked very well. I was interested in your script approach.
My next step (as well as sourcing a new head unit) is to set up another RPi Zero for Bluetooth. I was planning to do this without the GUI. I think there are a number of us interested in adding some install instructions for RPi in the user manual @WouterJD has created. I wonder if it would be helpful to have 3 sections such as (1) RPi baseline with ANT+ (2) RPi with Bluetooth and (3) RPi with Bluetooth without GUI (avoiding the wxPython issues). What do you think? I'd be happy to contribute/test etc.
Cheers, Keith

**Comment by wtal710174** [comment]
@MeanHat Could you check the script you are using? I have altered the run without sudo script in the comment above, since the bash command line options where lacking the quotation marks. Not sure what this could cause. Since I did this I did not get any more errors. If this does not help I need figure out how to exchange with PM's to help you troubleshoot.
Regarding the manual: @WouterJD asked me to open this thread to get the bluetooth installation on Raspberry mature. Once we have all the bugs out we can start writing the instruction for the manual. If I remember correctly this is done in Word. The basic ANT instruction by martin-vi can be found here: https://gist.github.com/martin-vi/5206b6b26faf6f2d6121f6c8f300e44d as you probably remember. This will need some updates, since it is not based on the latest source code and requirements. 
I am still testing some more features for the virtual environment as suggested by @switchabl like the installation of the node.js directly in the virtual environment. Once I am making progress there I will post. 
The other thing remaining is the issue with the crashing node.js when starting the bash script using my FortiusAntController.py this might get solved with the implementation of the node.js in the virtual environment.
Once all that is done and approved by WouterJD to be mature, I would gladly accept some help getting all this wrapped up and ready for WouterJD's manual.

**Comment by Lp64** [comment]
Hi All,
nice project. I excecuted all the commands from wtal710174 on a raspberry 4b like mikeherriman did. 
By executing the command  ./FortiusANT -b -g the GUI works. 
Clicking on Locate HW responds in: No (free) ANT-dongle found, Bluetooth interface available.
Terminal shows: 
17:36:05,374: Connected to Tacx Trainer T1932
17:36:05,678: Motor Brake Unit Firmware=6402 Serial=    0 year=2000 type=T190 Version2=0 MotorBrake=False

Clicking on Start in the GUI responds in: Connected to Tacx Trainer 1932
No (free) ANT-dongle found, Bluetooth interface opened to CTP

Terminal shows:
... requests.post() error HTTPConnectionPool(host='localhost', port=9999): Max retries exceeded with url: /ant (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xb002b050>: Failed to establish a new connection: [Errno 111] Connection refused'))

GUI is freezing. FortiusANT trainer is visible by bluetooth on a PC, but no chance to connect.
Any ideas; what is wrong? 
Ralf

**Comment by wtal710174** [comment]
@Lp64 As long as you see this error:
´´´
... requests.post() error HTTPConnectionPool(host='localhost', port=9999): Max retries exceeded with url: /ant (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xb002b050>: Failed to establish a new connection: [Errno 111] Connection refused'))
´´´
your node server is not running. Which is normal for a few seconds until node is up and running. To eliminate the GUI as possible source of the failure, start your FoertiusAnt with -b only and check if it works. After a view seconds you should see a similar line like during calibration. Any other terminal output you have posted is perfectly normal.

**Comment by MeanHat** [comment]
> @MeanHat Thanks for sharing this. I like the button idea. Looks very nice. Maybe I can find a Case with GPIO Connector that is big enough to fit this button. However I checked on the website and I think I like this switch even better:
> https://thepihut.com/products/rugged-metal-pushbutton-16mm-6v-rgb-momentary
> Check it out. It has a RBG LED ring, which allows you to show the status of the process, similar to what I am doing with the traffic light. Letting the RED LED ring blink in different ways could to the same.

Hi @wtal710174  - I like the RGB momentary pushbutton switch!
The script on the RPi3 is "FortiusANT Startup Script v0.2". I'm happy to try again. Working on an RPi Zero implementation currently.

**Comment by MeanHat** [comment]
@wtal710174 I have used this mini on/off shim with Bluetooth FortiusAnt on an RPi 3. It works fine and was easy to install and set up. It's not as elegant as your solution with traffic lights:  https://shop.pimoroni.com/products/onoff-shim. It then works headless using a systemd service to start up FortiusAnt to run with Zwift.

![PXL_20210117_111340162](https://user-images.githubusercontent.com/73789300/104838833-4bdc8480-58b5-11eb-84ae-1a61a5fd3619.jpg)

**Comment by trygvelu** [comment]
I did a zwift training session with the pi zero W, using an Ant dongle to communicate with my Genius trainer and BLE connection to the zwift computer. I had to download a node version for armv6 for the pi zero, as explained in the link in the Node JS section. I have a 7" LCD connected to the pi, but the wxPython gui does not work well for me. It freezes before I get to connect to the trainer. That got me thinking about alternative GUI options for embedded/headless solutions.
With node.js running (for BLE) it would maybe be feasible to implement a web based dashboard/frontend? There are many nice libraries available for building webpages with real-time data. 
Another option is tempting for me - build a pygame based dashboard, which can be run from the command line, and without running X11/wm in the background.

**Comment by wtal710174** [comment]
Thanks for your comment @trygvelu. Yes the wxPyhton GUI is quite unstable on the raspberry. Not sure why this is happening. I have been testing with a tkinter interface to use a 7" touch display to control the FortiusAnt application. As of now I am facing similar issues. Additionally the node server started by the FortiusAnt programq called using this interface, does not run stable at all. Therefore pygame could be another option to get a stable interface for the raspberry.  In case you come up with something working, please share. It might still have the issue with the unstable node server as long as you do not use it as a replacement for the GUI class in the FortiusAnt program itself.

**Comment by wtal710174** [comment]
### Node in virtual environment and updated startup script
I have been working on moving the node server into the virtual environment. Additionally I have moved the virtual environment into the FortiusAnt folder. This certainly does require an update of the script for starting FortiusAnt as well. In the following I have put a step by step instruction how to set this up.

#### Node in virtual environment
First you need to set up and activate a new virtual environment in the FortiusAnt folder
```
$ cd ~/FortiusANT/
$ virtualenv -p python3 --system-site-packages venv
$ source ~/FortiusANT/venv/bin/activate
```
Next you install wxPython (Download the [pre build package](https://drive.google.com/drive/folders/1owWkRTsvcnvq7ZSRc6zHS5dip2AMvW79?usp=sharing) and save to your Downloads directory in home)

```
$ pip install ~/Downloads/wxPython-4.1.1-cp37-cp37m-linux_armv7l.whl
```
Next install node to virtual environment
```
Download node 14.15.3
$ cd ~/Downloads/
$ wget https://nodejs.org/dist/v14.15.3/node-v14.15.3-linux-armv7l.tar.xz
# Untar downloaded archive
$ tar xf node-v14.15.3-linux-armv7l.tar.xz
# Go into folder with unzipped files
$ cd node-v14.15.3-linux-armv7l/
# Copy folder content to 
$ cp -R * ~/FortiusANT/venv/
```
and install the packages required for the FortiusANT node server
```
$ cd ~/FortiusANT/node/
$ npm install
```
Last step for node is to grant the node binary cap_net_raw privileges, so it can start/stop BLE advertising without root access.
```
$ sudo setcap cap_net_raw+eip $(eval readlink -f `which node`)
```
Next Install the FortiusANT dependencies in the virtual environment
```
$ pip install -r ~/FortiusANT/pythoncode/requirements.txt
```
To start FortiusANT add the following bash script to your home

```
$ cd ~
# Deactivate virtual environment
$ deactivate
$ touch FortiusANT.sh
# Open the new file with any editor you like, I prefer pico or nano
$ pico FortiusANT.sh
# Insert the following text including the !/bin/bash comment

#!/bin/bash

#FortiusANT Startup Script v0.3 running node in virtual environment

#Checking if script is running already
FILENAME=`basename $0`
pids=$(pgrep -x $FILENAME)
for pid in $pids
do
   if [ $pid -ne $$ ]; then
      echo "$FILENAME is already running. Exiting"
      exit 0
   fi
done

#Checking command line options

if [[ -z "$@" ]]
then
 #enter your default startup options here, if starting without commandline options ons
 #CLO="-a -b -g"
 CLO="-b"
else
 CLO=$@
fi

# Uncomment if you want automatic updates
# Execute 'git remote add upstream https://github.com/WouterJD/FortiusANT' in FortiusANT directory first before removing comment
#echo Getting FortiusANT updates from Github Repository
#git -C /home/pi/FortiusANT pull upstream master

sudo service bluetooth stop
echo Bluetooth Service stopped

source /home/pi/FortiusANT/venv/bin/activate
if [[ "$VIRTUAL_ENV" != "" ]]
then
  echo Virtual Environment set
else
  echo Virtual Environment not set
fi

hciconfig hci0 up
echo Bluetooth for BLE on internal adapter started

echo Starting FortiusANT ...
cd /home/pi/FortiusANT/pythoncode/
python FortiusAnt.py $CLO

deactivate
if [[ "$VIRTUAL_ENV" != "" ]]
then
  echo Virtual Environment set
else
  echo Virtual Environment not Set
fi

sudo service bluetooth start
echo Standard Bluetooth Service restarted
```
Save file with control+O and exit file with control+x and make script executable with the following command
```
$ chmod u+x FortiusANT.sh
```
If this is your first install you need to give the user access to the USB port.

#### Give user access to USB port and enable running without sudo 

Run the following commands to create an access file in /etc/udev/rules and reboot Raspberry to activate rule
```
$ sudo -i
$ cat << EOF > /etc/udev/rules.d/10-usbaccess.rules
# Allow users in group usbtacx to access usb
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", MODE="0664", GROUP="usbtacx"
EOF
# Create group usbtacx
$ addgroup usbtacx
# add user pi to this group
$ adduser pi usbtacx
# Reboot to activate rule
$ reboot now
```
and for running without sudo you need to use the following commands from this [previous post](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-756392335):

First give user the right to use hciconfig without sudo
```
$ sudo setcap 'cap_net_raw,cap_net_admin+eip' `which hciconfig`
```
Then enter the command 'sudo visudo' to edit the sudoers file. In the section
'# Allow members of group sudo to execute any command' add the following lines:
```
%usbtacx ALL=NOPASSWD: /bin/systemctl stop bluetooth.service
%usbtacx ALL=NOPASSWD: /bin/systemctl start bluetooth.service
```
This will grant all member of the group usbtacx the right to start bluetooth service without entering a password.
Now you can use sudo service bluetooth stop and sudo service bluetooth start as user pi without entering a password. 

#### Let's fire it up
Connect your trainer to one of the USB ports of the Raspberry Pi and execute FortiusANT with
```
# -b for command line bluetooth startup, add any command line parameter as required, like -g to start with GUI
$ ~/FortiusANT.sh -b
```
Next, the calibration procedure will start with a pedal kick. After the procedure is complete, the bluetooth starts broadcasting identifying the Raspberry Pi as periphal device with the identifier 'FortiusANT Trainer'. This one you can connect to your CTP.

After the calibration you might see a couple of error messages. This happens while the http server is being established and the program is not able to access it. After 5-10 secondes the connection should be stable and you will see the normal output in the terminal like the one below.
```
20:08:31,698: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:32,700: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:33,702: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:34,703: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:35,705: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
20:08:36,706: Target=100W Speed= 0.0kmh hr=  0 Current=  0W Cad=  0 r=   0 T= 15
```
Let me know if this works for you, could be that I forgot to add some steps. Difficult to write this without a clean system. 

You can easily start this script automatically by adding the following command line to the end of your .bashrc file
```
# edit with 
$ pico ~/.bashrc
# add the following line
/home/pi/FortiusANT.sh &
```
This will start the FortiusANT every time you start your Raspberry. Opening a terminal will not start it again. All you need to do is kick a pedal about 1 minute after you started the raspberry. 
I expect to post an update on the traffic light visual control soon, which will help to visualize the status of the FortiusAnt program.

**Comment by wtal710174** [comment]
### Visual Status Indicator for FortiusANT
As stated before, I still can not get the script in the post [FortiusAnt One Button Controller for Raspberry](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-757495938) to work reliably, since the node server always stops working after latest 10 minutes.
Therefore I had the idea to use the traffic light indicator at least to display the status of FortiusAnt, while not using it to launch FortiusAnt. Currently I am starting FortiusAnt using my FortiusANT.sh script for running the program with node in the virtual environment and I am launching the Status Indicator program in the same way. So I have 2 programs running independently. The Status indicator does not even run in a virtual environment. Both are being launched in the .bashrc file.

The Status indicator shows red if nothing is running, yellow (blinking), if FortiusAnt is running and node has not started the http server for broadcasting and green once everything is up and running. Additionally a long press (>3s) of the push button shuts the raspberry down. The light status changes, if anything stops working at some point.

Currently this seems to work fine. Below you find the source code attached.
 
[FortiusAntStatusIndicator.py.zip](https://github.com/WouterJD/FortiusANT/files/5860549/FortiusAntStatusIndicator.py.zip)

**Comment by povey76** [comment]
I have tried a couple of times now with updated link (reformatting and starting again) and I am getting an error when it comes to: 
$ cp -R * ~/FortiusANT/venv/
The error reads cp: target 'home/pi/fortiusant/venv/' is not a directory? Every thing has worked up until this point.

I have looked in the files and the file structure has fortiusant and within that node and venv are present.                                                                                                                                

Any suggestions would be appreciated, took some time to get to this point - very new at this!

**Comment by wtal710174** [comment]
@povey76: Difficult to diagnose this based on your description. The only thing that seems a possible cause is the fact that Linux is case sensitive. So if you have named your folder "FortiusANT" and you use "fortiusant" in your command it will not work, because this is using different names for the directory. Hope this helps, otherwise we have to try to troubleshoot some more in detail.

**Comment by povey76** [comment]
> @povey76: Difficult to diagnose this based on your description. The only thing that seems a possible cause is the fact that Linux is case sensitive. So if you have named your folder "FortiusANT" and you use "fortiusant" in your command it will not work, because this is using different names for the directory. Hope this helps, otherwise we have to try to troubleshoot some more in detail.

Thanks, obvious really!! Amended and all good to try, although:

18:04:26,084: Dongle - No expected replay from dongle, Bluetooth interface available?

Anything that I need to apply for the Garmin ANT usb to be recognised?

**Comment by wtal710174** [comment]
@povey76: If you do not want to use the built in bluetooth while connecting the Tacx Head Unit using USB you need to refer to the Instruction that @martin-vi posted on his Git here: [Ant Linux Instruction](https://gist.github.com/martin-vi/5206b6b26faf6f2d6121f6c8f300e44d). There you will find the rights you need to set for the Ant Dongle. Additionally you need to make sure that you start the FortiusANT script without the -b option for bluetooth.

**Comment by povey76** [comment]
> @povey76: If you do not want to use the built in bluetooth while connecting the Tacx Head Unit using USB you need to refer to the Instruction that @martin-vi posted on his Git here: [Ant Linux Instruction](https://gist.github.com/martin-vi/5206b6b26faf6f2d6121f6c8f300e44d). There you will find the rights you need to set for the Ant Dongle. Additionally you need to make sure that you start the FortiusANT script without the -b option for bluetooth.

Ok, so I don't need the ANT dongle if using this method - even better! So as it is, I can connect the usb from the fortuis and when peddling/calibration I should see the response on the Screen (started with -g) and will transmit via bluetooth to Apple TV (zwift)? Appreciate the support- if this works later I'll be made up! Not doubting code, my newness to raspberry and code!!😂

**Comment by wtal710174** [comment]
Yes, this is the way it should work. Start with -b -g, if you want the GUI Interface. -b is sufficient if you are OK with just the output in the terminal.
Good look and have a great first ride!

**Comment by povey76** [comment]
> Yes, this is the way it should work. Start with -b -g, if you want the GUI Interface. -b is sufficient if you are OK with just the output in the terminal.
> 
> Good look and have a great first ride!

Thanks, when kids are in bed I'll be testing!! Is there an addition that you don't have to enter a command code? I have a spare screen so may mount that (looks good), but would be ideal to start peddling or hit a button to switch off and kill when done? I've seen some comments on this but still a novice to grasp where it needs to to copied and pasted!! Sorry, and really appreciate the help & support - what a group!!👍👍👊

**Comment by wtal710174** [comment]
There is several ways, for the beginning I would use the screen and observe the output after starting the program manually in a terminal. This way you will understand if the Bluetooth is connected, and the node server is running and you will see all the possible failures that might happen. It is much simpler to troubleshoot by posting the output and errors you get. 
Regarding the start without interaction, there are several ways of making the unit headless. Two are described in this issue here.

[Traffic Light](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-757495938) which I am using with the script in this [link](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-766138122).

@MeanHat has a pretty nice and simple solution with his [on/off shim](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-761776533).

The best solution for the future is probably going to be the [pull request](https://github.com/WouterJD/FortiusANT/pull/232) by @decodeais, which is for ANT systems, however can be used for the bluetooth solution as well, with some minor modifications.

So lots of possibilities and things to learn.

**Comment by povey76** [comment]
> There is several ways, for the beginning I would use the screen and observe the output after starting the program manually in a terminal. This way you will understand if the Bluetooth is connected, and the node server is running and you will see all the possible failures that might happen. It is much simpler to troubleshoot by posting the output and errors you get. 
> 
> Regarding the start without interaction, there are several ways of making the unit headless. Two are described in this issue here.
> 
> 
> 
> [Traffic Light](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-757495938) which I am using with the script in this [link](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-766138122).
> 
> 
> 
> @MeanHat has a pretty nice and simple solution with his [on/off shim](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-761776533).
> 
> 
> 
> The best solution for the future is probably going to be the [pull request](https://github.com/WouterJD/FortiusANT/pull/232) by @decodeais, which is for ANT systems, however can be used for the bluetooth solution as well, with some minor modifications.
> 
> 
> 
> So lots of possibilities and things to learn.
> 
> 

Good new's is that she is alive and communicating with zwift! Although I have some errors as picture? What's required?![image](https://user-images.githubusercontent.com/78272870/107580836-19d9ec00-6bef-11eb-88cc-6e1d73b0984a.jpeg)

**Comment by wtal710174** [comment]
> Good new's is that she is alive and communicating with zwift! Although I have some errors as picture? What's required?![image](https://user-images.githubusercontent.com/78272870/107580836-19d9ec00-6bef-11eb-88cc-6e1d73b0984a.jpeg)

Not really sure, what this can be. I guess the USB cable is connected correctly. Maybe try another cable. If this does not work, check if you have given the User the right to access the USB port. Not sure if this can be the issue since FortiusAnt is finding the hardware in the first lines. 

Let me know how this works. If it does not the next step would be to run FortiusAnt in debug mode and post the logfile.

**Comment by switchabl** [comment]
@wtal710174 This seems to be the same problem as #251 and #253, so it probably isn't something that is wrong at your end.

**Comment by povey76** [comment]
> @wtal710174 This seems to be the same problem as #251 and #253, so it probably isn't something that is wrong at your end.

Thanks for reply, seems to communicate ok - so not concerned - just wanted to check it wasn't an installation issue!

**Comment by WouterJD** [comment]
@wtal710174 great instruction-set; RPI4 is working in console-mode.
Next step is BLE and then wxPython.

Please explain, what is the reason to use the virtual environment.

**Comment by wtal710174** [comment]
@WouterJD The reason for the virtual environment is that python can not differentiate different package version if you are running more projects on your Raspi. In the virtual environment it will have the package versions it requires for the project. It is more or less just good practice in python to handle this issue. If your project is going to be the only python project on the raspi you can skip the virtual environment and install and run directly. A good excessive explanation you can find here: [Python Virtual Environments: A Primer](https://realpython.com/python-virtual-environments-a-primer/)

**Comment by WouterJD** [comment]
Many thanks, I expected this but wondered if it had to do with BLE. Thanks to your support my RPI is now working well. Next are: BLE and wxPython. Moving forward😉

**Comment by WouterJD** [comment]
> * Non Gui version:  Yes this is a great idea.

There is a UseGui constant in constants.py, which was not tested (by me) untill now.
In FortiusANT 5.3 this variable will be set to False when wx is not available, and this requirement is met then.
-g is then not available either.

**Comment by WouterJD** [comment]
> #echo Getting FortiusANT updates from Github Repository
> #git -C /home/pi/FortiusANT pull upstream master

Automatc updating is NOT implemented by design; at the moment of implementation this check it was requested to inform, not to update.
I agreed with that requirement: it avoids that a working system is "improved" with new errors.

**Comment by wtal710174** [comment]
@WouterJD for wxpython just install the pre build package. It will work on the RPI4. 
BTW: I am having some issues with the headless version and the calibration. If I fire up the RPI and immediatly calibrate as soon as I can, the calibration fails (Runs until timeout). If I wait some time and calibrate then the calibration works. Not sure why, would be intersting if I am the only one that is experiencing it.
Nice idea checking the presence of the wx package and disabling gui automatically. Maybe still accept the -g option, but starting without gui with a message in the console, instead of stopping the execution, because of an invalid option. Hope I understood this correctly, did not check / notice the version 5.3 code yet.

**Comment by WouterJD** [comment]
> BTW: I am having some issues with the headless version and the calibration. If I fire up the RPI and immediatly calibrate as soon as I can, the calibration fails (Runs until timeout).

Perhaps RPI not fully booted yet. Add sleep in script and see what happens.
Timing will not be revealed by logfile

**Comment by WouterJD** [comment]
> Maybe still accept the -g option

It's how it is. If you don't want gui, don't specify -g flag. If you do: error message. Easy😉

**Comment by MeanHat** [comment]
Hi @wtal710174 - you asked for other people's experience with calibration on headless RPi. I'm using RPi3 and calibration has never timed out in probably 50+ starts. If you need data on timing please let me know and I can check the log file.

**Comment by wtal710174** [comment]
@MeanHat Yes that would be very helpful. Thanks for the offer. Just trying to get a feeling about what is "normal" for calibration time.

**Comment by wtal710174** [comment]
Just for reference: Calibration takes ~100-120s if it works when starting up. Trainer is in my basement at roughly 19°C.

**Comment by wtal710174** [comment]
### Update on the one Button Controller
Upgraded my [temporary solution](https://github.com/WouterJD/FortiusANT/issues/192#issuecomment-757495938) to a boxed one with 5 LED's. 3 LED's as before and a white and blue one to indicate e.g. steering calibration commands or ANT+ signals controls. If I have some time I will create a fork from #232 to implement the LED's directly into the FortiusANT Program. Inside is not very professional but good enough. See attached pictures.

![IMG_7593(1)](https://user-images.githubusercontent.com/62606555/109390006-5f94f500-790f-11eb-9300-9cc9d9a126bc.JPG)
![IMG_7594](https://user-images.githubusercontent.com/62606555/109390007-615eb880-790f-11eb-883f-fc3ce7c0bf04.JPG)
![IMG_7595](https://user-images.githubusercontent.com/62606555/109390009-628fe580-790f-11eb-9867-5b633940630e.JPG)
![IMG_7596](https://user-images.githubusercontent.com/62606555/109390010-63287c00-790f-11eb-8c59-2f7359ca6805.JPG)

**Comment by MeanHat** [comment]
> @MeanHat Yes that would be very helpful. Thanks for the offer. Just trying to get a feeling about what is "normal" for calibration time.

Hi @wtal710174 - I have measured the calibration and the times range from 31 to 52 seconds. Hope that helps. Trainer is in room with temperature around 17c.

Your switch and lights are looking good!

**Comment by MeanHat** [comment]
Hi again @wtal710174  - I thought you might like to see my headless RPi3 with mini TFT display. I have credited @WouterJD's FortiusANT.
![PXL_20210227_191401553 MP](https://user-images.githubusercontent.com/73789300/109397955-882de680-7931-11eb-8fd8-c7bb56d014cf.jpg)
![PXL_20210227_184842052](https://user-images.githubusercontent.com/73789300/109397958-8cf29a80-7931-11eb-9f07-5109aa3d19da.jpg)

**Comment by WouterJD** [comment]
Hi guys! Great work and I sure appreciate it.
My RPI4 is now working with BLE :-)

Did any of you already decide where to start FortiusAnt (and how), keeping a console "visible", assuming a monitor would be connected:
- rc.local
- /etc/xdg/autostart/
- other??

When starting from rc.local, it sure boots but has no visible stdout...

**Comment by WouterJD** [comment]
> I thought you might like to see my headless RPi3 with mini TFT display

I sure like it!!
Would you share some details?

**Comment by wtal710174** [comment]
The display looks great. Would be very interested in seeing the details. I guess you adapted the code as well. Could you share? Even if the Display would be too small for me to read 🤣. This is why I diverted to the LED thing.

**Comment by MeanHat** [comment]
> Hi guys! Great work and I sure appreciate it.
> My RPI4 is now working with BLE :-)
> 
> Did any of you already decide where to start FortiusAnt (and how), keeping a console "visible", assuming a monitor would be connected:
> 
> * rc.local
> * /etc/xdg/autostart/
> * other??
> 
> When starting from rc.local, it sure boots but has no visible stdout...

Hi @WouterJD - I autostart FortiusANT with a headless RPi using /.bashrc. This starts for FortiusANT Controller which in turn starts FortiusAnt.py through a .sh shell script (based on the one from @wtal710174 above in this issue: #192)

**Comment by MeanHat** [comment]
> > I thought you might like to see my headless RPi3 with mini TFT display
> 
> I sure like it!!
> Would you share some details?

HI @WouterJD - I have uploaded the code to https://github.com/MeanHat/FortiusANT_controller. I'm new to Python so it is certainly not elegant or fool proof. I'm very open to corrections and suggestions for improvement. Hope this is useful. The display is relatively cheap and although the text is small is has two buttons which I use to start-up and close-down the RPi. Good luck with your RPi project.

**Comment by MeanHat** [comment]
> The display looks great. Would be very interested in seeing the details. I guess you adapted the code as well. Could you share? Even if the Display would be too small for me to read 🤣. This is why I diverted to the LED thing.

HI @wtal710174 - many thanks for your  instructions above on running FortiusANT using Bluetooth on an RPi. I've successfully used it on two RPi3s now- one driving an iMagic and this one with the display driving a Tacx Fortius. Great work!

I've uploaded the code for the miniTFT controller to https://github.com/MeanHat/FortiusANT_controller. As I mentioned to @WouterJD I'm new to Python so very open to corrections and suggestions for improvement.

Your comment about the size of the text is noted (I know what you mean) - it's obviously possible to just display colours on the screen (https://github.com/adafruit/Adafruit_CircuitPython_RGB_Display/blob/master/examples/rgb_display_minipitfttest.py) which would be an alternative to the text or LEDs.

Thanks once again for enabling me to get Bluetooth running

**Comment by decodeais** [comment]
Hi @WouterJD I have read that you are running BLE an the RPI4. I tried it too, but with no success. Did you use the internal Bluetooth or did you use a stick ?
I tested it with Rouvy, sometimes I saw that the FortiusAnt device is working, but I could not connect. I have no Idea where to search for the reason.

**Comment by WouterJD** [comment]
@decodeais I followed the installation instructions #192 and it worked right-away.
Do not forget next two commands:
```
sudo service bluetooth stop
sudo hciconfig hci0 up
```
I understand this is to "disable standard ble and enable ble for nodejs" but that's my interpretation.

I'm gathering all exchanged data into the manual, the publication will take some time.

**Comment by decodeais** [comment]
Thank you, I already did everything written in the description. When you said that it is really working with the internaI RP4 bluetooth i started testing again.
My tablet stuck in the connection screen while the RPI4 was nearly frozen.
So I looked for another device to test. I Installed Rouvy AR on my smartphone and it works, but the display is a little to small. Has anyone an idea what is wrong with my tablet (Galaxy Tab A with Android 8.1.0)

**Comment by MeanHat** [comment]
@WouterJD - do you need/want any help updating the manual for RPi installation? Happy to lend a hand.

**Comment by WouterJD** [comment]
When giving the command

`$ sudo pip3 install wxPython-4.1.1-cp37-cp37m-linux_armv7l.whl`

I get

```
Requirement 'wxPython-4.1.1-cp37-cp37m-linux_armv7l.whl' looks like a filename, but the file does not exist
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Processing ./wxPython-4.1.1-cp37-cp37m-linux_armv7l.whl
Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: '/home/pi/FortiusAnt/raspberry/wxPython-4.1.1-cp37-cp37m-linux_armv7l.whl'

```
Any suggestions?

**Comment by WouterJD** [comment]
> do you need/want any help updating the manual for RPi installation

Thanks @MeanHat I'll come back to it when required.
Currently scripting the whole installation (after some bootstrap); manual page is next step.

**Comment by WouterJD** [comment]
@wtal710174, @decodeais, @MeanHat 

Thanks for all the inspiration and documentation!
![image](https://user-images.githubusercontent.com/47486941/109671662-915ac580-7b74-11eb-8089-e936fee113ad.png)

I will make a wiki-page where all special pictures will be gathered "Raspberry page of glory"; who has a fancier title for the page?

**Comment by WouterJD** [comment]
Please enjoy https://github.com/WouterJD/FortiusANT/wiki/Raspberry-page-of-glory

**Comment by WouterJD** [comment]
I have integrated Raspberry Pi into FortiusAnt, code, installation scripts and manual.
Branch Raspberry_Release6 is checked in; review is appreciated.

Tonight I cleaned the SD-card and reinstalled everything.
When the Raspberry is started, FortiusAnt is started, so far so good.

BUT the bluetooth hci0 device does not work:
```
pi@TacxFortiusPi:~ $ sudo service bluetooth stop
pi@TacxFortiusPi:~ $ sudo hciconfig hci0 up
Can't init device hci0: Connection timed out (110)
```
and when I give the sudo service bluetooth status:
```
* bluetooth.service - Bluetooth service
   Loaded: loaded (/lib/systemd/system/bluetooth.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2021-03-05 00:24:26 CET; 14s ago
     Docs: man:bluetoothd(8)
 Main PID: 2346 (bluetoothd)
   Status: "Running"
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/bluetooth.service
           `-2346 /usr/lib/bluetooth/bluetoothd

Mar 05 00:24:26 TacxFortiusPi systemd[1]: Starting Bluetooth service...
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: Bluetooth daemon 5.50
Mar 05 00:24:26 TacxFortiusPi systemd[1]: Started Bluetooth service.
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: Starting SDP server
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: Bluetooth management interface 1.18 initialized
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: Sap driver initialization failed.
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: sap-server: Operation not permitted (1)
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: Endpoint registered: sender=:1.23 path=/MediaEndpoint/A2DPSource
Mar 05 00:24:26 TacxFortiusPi bluetoothd[2346]: Endpoint registered: sender=:1.23 path=/MediaEndpoint/A2DPSink
Mar 05 00:24:28 TacxFortiusPi bluetoothd[2346]: Loading LTKs timed out for hci0

```

Suggestions appreciated!

**Comment by MeanHat** [comment]
Hi @WouterJD - great work on the new Release and the updated manual https://github.com/WouterJD/FortiusANT/blob/Raspberry_Release6/doc/FortiusANTUserManual.pdf. I'll happily test out the new v6. 

I've changed the code running the miniTFT display as Bluetooth was dropping out after about 2 mins. I will post the updated version which fixes this.

Hope you are able to fix your Bluetooth issue - if I have any suggestions I'll let you know.

**Comment by WouterJD** [comment]
> BUT the bluetooth hci0 device does not work:
> 
> ```
> pi@TacxFortiusPi:~ $ sudo service bluetooth stop
> pi@TacxFortiusPi:~ $ sudo hciconfig hci0 up
> Can't init device hci0: Connection timed out
> ```

Yesterday (ahum it was 01:00) I switched the RPI off, left it for tonite to redo the installation.
I just switched on - issue gone and Bluetooth working. Hum hum. But for now problem solved.

**Comment by WouterJD** [comment]
@wtal710174 I hope hope you can live with how I scripted and documented. After all it's your work I "scrambled". It sure gives thoughts how to redo the windows installation.

**Comment by WouterJD** [comment]
> changed the code running the miniTFT display

@MeanHat keep us informed in #237 👍

**Comment by wtal710174** [comment]
> 
> 
> @wtal710174 I hope hope you can live with how I scripted and documented. After all it's your work I "scrambled". It sure gives thoughts how to redo the windows installation.

Makes me happy to see that I could give something back to this awesome project, which increased my time on the bike tremendously during the winter 🥇. I will check the documentation and let you know if I have any remarks.

**Comment by wtal710174** [comment]
> 
> Yesterday (ahum it was 01:00) I switched the RPI off, left it for tonite to redo the installation.
> I just switched on - issue gone and Bluetooth working. Hum hum. But for now problem solved.

And I thought stuff like this happens on Windows only ...

**Comment by WouterJD** [comment]
> I thought stuff like this happens on Windows only

A hard wake-up call :-(.

**Comment by WouterJD** [comment]
@martin-vi do you follow this issue?

@wtal710174 references https://gist.github.com/martin-vi/5206b6b26faf6f2d6121f6c8f300e44d and I think it would be good to integrate that into the current description and manual.

My consideration to see BLE as the default for Raspberry Pi is that it's default onboard AND there is a lot of confusion on ANT-dongles. The Raspberry Pi can then be used from a Smartphone. (Ahum ... this part is not in the manual yet).

What is your consideration to use ANT?

**Comment by WouterJD** [comment]
@MeanHat do you use ANT or BLE?
Would it be possible to script the ANT-configuration, even when some modifications are needed?

**Comment by WouterJD** [comment]
I intend to extend the manual with the following section:

### 3.12.7	Now pair with a CTP
Zwift, Trainer Road, Rouvy, Sufferfest, Golden Cheetah, Road Grand Tour, TTS and many others can now pair with the Bluetooth trainer “FortiusANT Trainer” on a device that supports BLE. Perhaps a Smartphone is the most generally available device to use and from there the screen can be streamed to a TV. Of course, you can also use a PC with a Bluetooth dongle.
I would think that this the recommended configuration because BLE is native on Raspberry Pi, a Bluetooth dongle is usually cheaper than an ANT-dongle and because ANT-dongles are hard to get (see issue #61). Also, it seems that the market of CTP’s is moving from ANT to BLE; usually Bluetooth is preferred and/or supported before ANT.
- - -
Comment appreciated

**Comment by WouterJD** [comment]
Another question: when the .sh files are fetched from git, the execute bit is not set and hence the files are nmot executable using a double-click from the file-manager. How could that be solved?

**Comment by MeanHat** [comment]
> @martin-vi do you follow this issue?
> 
> @wtal710174 references https://gist.github.com/martin-vi/5206b6b26faf6f2d6121f6c8f300e44d and I think it would be good to integrate that into the current description and manual.
> 
> My consideration to see BLE as the default for Raspberry Pi is that it's default onboard AND there is a lot of confusion on ANT-dongles. The Raspberry Pi can then be used from a Smartphone. (Ahum ... this part is not in the manual yet).
> 
> What is your consideration to use ANT?

Supporting @martin-vi's points about both Bluetooth and phones I'd add:
- using Bluetooth was great for users running Zwift on an Android phone as Zwift did not support FE-C with ANT though this was fixed in Dec 20 (https://forums.zwift.com/t/zwiftgame-android-faq-v-1-0-60239/520294) which now gives users a choice.
- ANT runs successfully on an RPi0 which is very cost-effective (other than the cost of the dongles of course). I've not yet managed to get Bluetooth running on an RPi0 but will keep experimenting with a spare Zero as RPi0 + onboard Bluetooth would be really cost-effective

**Comment by MeanHat** [comment]
> @MeanHat do you use ANT or BLE?
> Would it be possible to script the ANT-configuration, even when some modifications are needed?

Hi @WouterJD - I've been using Bluetooth with the minTFT display but have used ANT for a headless RPi0 version I did for my neighbour last year. I'll do that. 
It's a great suggestion to add ANT to the miniTFT controller. I'll need to generate a log file for an RPi running with ANT (unless you have one to hand you could share) to set this up and test it. I can install FortiusANT on a spare RPi0 and get a log file (based on -d1 or higher combination parameter). I'll post the updated version of the miniTFT controller that fixes the Bluetooth issue first (it's complete but being tested) and then do the ANT version.

**Comment by MeanHat** [comment]
I've seen that also. Double clicking the .sh file from the file-manager
provides me with three main options:

   1. Execute
   2. Execute in Terminal Mode
   3. Open (for editing)

When testing FortiusANT with a GUI I found it preferable to start using
option #2. Running headless I have called the .sh file from a startup
script which works fine (I have also opened it from the python miniTFT
Controller using Popen).

Not sure if that helps?

**Comment by decodeais** [comment]
hello @WouterJD, in Windows you can set the execution flag by this git  command.
`git update-index --chmod=+x <filename>`
You can check the result by `git ls-files --stage`

**Comment by MeanHat** [comment]
Extra section looks good. I think it would be useful to highlight that ANT
also works (although not the preferred option) in case someone is
specifically looking for that option. For example:

   - Zwift, Trainer Road, Rouvy, Sufferfest, Golden Cheetah, Road Grand
   Tour, TTS and many others can now pair with the Bluetooth trainer*
   identified as* “FortiusANT Trainer” on a device that supports BLE.
   Perhaps a Smartphone is the most generally available device to use and from
   there the screen can be streamed to a TV* or monitor*. Of course, you
   can also use a PC with a Bluetooth dongle. *It is also possible to use
   ANT as an alternative to BLE though this is not generally preferred by
   FortiusANT users as the ANT dongles are often more expensive and harder to
   buy, and most CTPs seem to be committed to supporting BLE.*

Just a suggestion...

**Comment by decodeais** [comment]
Would it be possible to make kind of list which Tablet or smartphone is working. For example:
My Samsung Tablet A6 works with internal ANT+ but  not with BLE. 
My Smartphone S5  has internal ANT+ too,  but Android 10 has a bug and the internal ANT+ does not work. Fortunately  BLE is working here.
It would be bad when all the  Problems of the tablets and smartphones has to be discussed here separately.

**Comment by WouterJD** [comment]
A hardware compatibility list would be very limitted since usually when things work, it's not reported.
Would become a known issues list

**Comment by WouterJD** [comment]
Discussion moved to #274, I suggest we close this one.
We always reopen if required
